File: packages/autonomo-python/autonomo/registry.py
"""
Element Registry - Tracks interactive elements for AI control

Components register themselves when mounted, unregister when unmounted.
This provides a live map of what the AI can interact with.
"""

from enum import Enum
from dataclasses import dataclass, field
from typing import Callable, Optional, Dict, Any, List
import re


class ElementType(Enum):
    BUTTON = "button"
    INPUT = "input"
    TOGGLE = "toggle"
    SELECT = "select"
    LINK = "link"
    CUSTOM = "custom"


@dataclass
class ElementHandler:
    """Handler for an interactive element"""
    type: ElementType
    handler: Callable[[Optional[str]], None]
    disabled: bool = False
    get_value: Optional[Callable[[], str]] = None
    on_submit: Optional[Callable[[], None]] = None
    hint: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None


@dataclass
class ElementInfo:
    """Information about a registered element"""
    id: str
    type: ElementType
    disabled: bool = False
    value: Optional[str] = None
    hint: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        result = {
            "id": self.id,
            "type": self.type.value,
        }
        if self.disabled:
            result["disabled"] = self.disabled
        if self.value is not None:
            result["value"] = self.value
        if self.hint is not None:
            result["hint"] = self.hint
        if self.meta is not None:
            result["meta"] = self.meta
        return result


class ElementRegistry:
    """Singleton registry for all interactive elements"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._elements: Dict[str, ElementHandler] = {}
            cls._instance._listeners: List[Callable[[], None]] = []
        return cls._instance
    
    def register(self, id: str, handler: ElementHandler) -> Callable[[], None]:
        """Register an interactive element. Returns unregister function."""
        self._elements[id] = handler
        self._notify_change()
        return lambda: self.unregister(id)
    
    def unregister(self, id: str) -> None:
        """Unregister an element"""
        if id in self._elements:
            del self._elements[id]
            self._notify_change()
    
    def get(self, id: str) -> Optional[ElementHandler]:
        """Get handler for an element"""
        return self._elements.get(id)
    
    def has(self, id: str) -> bool:
        """Check if element exists"""
        return id in self._elements
    
    def list(self) -> List[str]:
        """List all element IDs"""
        return list(self._elements.keys())
    
    def get_all(self) -> List[ElementInfo]:
        """Get detailed info for all elements"""
        result = []
        for id, handler in self._elements.items():
            value = handler.get_value() if handler.get_value else None
            result.append(ElementInfo(
                id=id,
                type=handler.type,
                disabled=handler.disabled,
                value=value,
                hint=handler.hint,
                meta=handler.meta,
            ))
        return result
    
    def find(self, pattern: str) -> List[ElementInfo]:
        """Find elements matching a regex pattern"""
        regex = re.compile(pattern)
        return [el for el in self.get_all() if regex.search(el.id)]
    
    def clear(self) -> None:
        """Clear all elements"""
        self._elements.clear()
        self._notify_change()
    
    @property
    def size(self) -> int:
        """Get count of registered elements"""
        return len(self._elements)
    
    def on_change(self, listener: Callable[[], None]) -> Callable[[], None]:
        """Subscribe to registry changes"""
        self._listeners.append(listener)
        return lambda: self._listeners.remove(listener)
    
    def _notify_change(self) -> None:
        for listener in self._listeners:
            listener()


# Singleton instance
registry = ElementRegistry()


def register_tap_handler(
    id: str,
    handler: Callable[[], None],
    *,
    disabled: bool = False,
    hint: Optional[str] = None,
    meta: Optional[Dict[str, Any]] = None,
) -> Callable[[], None]:
    """Register a tap handler for a component"""
    return registry.register(id, ElementHandler(
        type=ElementType.BUTTON,
        handler=lambda _: handler(),
        disabled=disabled,
        hint=hint,
        meta=meta,
    ))


def register_fill_handler(
    id: str,
    handler: Callable[[str], None],
    *,
    get_value: Optional[Callable[[], str]] = None,
    on_submit: Optional[Callable[[], None]] = None,
    disabled: bool = False,
    hint: Optional[str] = None,
    meta: Optional[Dict[str, Any]] = None,
) -> Callable[[], None]:
    """Register a fill handler for an input"""
    return registry.register(id, ElementHandler(
        type=ElementType.INPUT,
        handler=lambda v: handler(v or ""),
        get_value=get_value,
        on_submit=on_submit,
        disabled=disabled,
        hint=hint,
        meta=meta,
    ))


def register_toggle_handler(
    id: str,
    handler: Callable[[Optional[str]], None],
    *,
    get_value: Optional[Callable[[], str]] = None,
    disabled: bool = False,
    hint: Optional[str] = None,
    meta: Optional[Dict[str, Any]] = None,
) -> Callable[[], None]:
    """Register a toggle handler for a switch/checkbox"""
    return registry.register(id, ElementHandler(
        type=ElementType.TOGGLE,
        handler=handler,
        get_value=get_value,
        disabled=disabled,
        hint=hint,
        meta=meta,
    ))

File: packages/autonomo_flutter/lib/src/registry.dart
/// Element Registry - Tracks interactive elements for AI control
///
/// Widgets register themselves when mounted, unregister when disposed.
/// This provides a live map of what the AI can interact with.

/// Type of interactive element
enum ElementType {
  button,
  input,
  toggle,
  select,
  link,
  custom,
}

/// Handler for an interactive element
class ElementHandler {
  final ElementType type;
  final Future<void> Function([String? value]) handler;
  final bool disabled;
  final String? Function()? getValue;
  final Future<void> Function()? onSubmit;
  final String? hint;
  final Map<String, dynamic>? meta;

  const ElementHandler({
    required this.type,
    required this.handler,
    this.disabled = false,
    this.getValue,
    this.onSubmit,
    this.hint,
    this.meta,
  });
}

/// Information about a registered element
class ElementInfo {
  final String id;
  final ElementType type;
  final bool disabled;
  final String? value;
  final String? hint;
  final Map<String, dynamic>? meta;

  const ElementInfo({
    required this.id,
    required this.type,
    this.disabled = false,
    this.value,
    this.hint,
    this.meta,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.name,
        if (disabled) 'disabled': disabled,
        if (value != null) 'value': value,
        if (hint != null) 'hint': hint,
        if (meta != null) 'meta': meta,
      };
}

typedef RegistryChangeListener = void Function();

/// Singleton registry for all interactive elements
class ElementRegistry {
  static final ElementRegistry _instance = ElementRegistry._internal();
  factory ElementRegistry() => _instance;
  ElementRegistry._internal();

  final Map<String, ElementHandler> _elements = {};
  final Set<RegistryChangeListener> _listeners = {};

  /// Register an interactive element
  /// Returns a function to unregister
  void Function() register(String id, ElementHandler handler) {
    _elements[id] = handler;
    _notifyChange();
    return () => unregister(id);
  }

  /// Unregister an element
  void unregister(String id) {
    _elements.remove(id);
    _notifyChange();
  }

  /// Get handler for an element
  ElementHandler? get(String id) => _elements[id];

  /// Check if element exists
  bool has(String id) => _elements.containsKey(id);

  /// List all element IDs
  List<String> list() => _elements.keys.toList();

  /// Get detailed info for all elements
  List<ElementInfo> getAll() {
    return _elements.entries.map((entry) {
      final handler = entry.value;
      return ElementInfo(
        id: entry.key,
        type: handler.type,
        disabled: handler.disabled,
        value: handler.getValue?.call(),
        hint: handler.hint,
        meta: handler.meta,
      );
    }).toList();
  }

  /// Find elements matching a pattern
  List<ElementInfo> find(RegExp pattern) {
    return getAll().where((el) => pattern.hasMatch(el.id)).toList();
  }

  /// Clear all elements
  void clear() {
    _elements.clear();
    _notifyChange();
  }

  /// Get count of registered elements
  int get size => _elements.length;

  /// Subscribe to registry changes
  void Function() onChange(RegistryChangeListener listener) {
    _listeners.add(listener);
    return () => _listeners.remove(listener);
  }

  void _notifyChange() {
    for (final listener in _listeners) {
      listener();
    }
  }
}

/// Global registry instance
final registry = ElementRegistry();

/// Register a tap handler for a widget
void Function() registerTapHandler(
  String id,
  Future<void> Function() handler, {
  bool disabled = false,
  String? hint,
  Map<String, dynamic>? meta,
}) {
  return registry.register(
    id,
    ElementHandler(
      type: ElementType.button,
      handler: ([_]) => handler(),
      disabled: disabled,
      hint: hint,
      meta: meta,
    ),
  );
}

/// Register a fill handler for a text input
void Function() registerFillHandler(
  String id,
  Future<void> Function(String value) handler, {
  String? Function()? getValue,
  Future<void> Function()? onSubmit,
  bool disabled = false,
  String? hint,
  Map<String, dynamic>? meta,
}) {
  return registry.register(
    id,
    ElementHandler(
      type: ElementType.input,
      handler: ([value]) => handler(value ?? ''),
      getValue: getValue,
      onSubmit: onSubmit,
      disabled: disabled,
      hint: hint,
      meta: meta,
    ),
  );
}

/// Register a toggle handler for a switch/checkbox
void Function() registerToggleHandler(
  String id,
  Future<void> Function([String? value]) handler, {
  String? Function()? getValue,
  bool disabled = false,
  String? hint,
  Map<String, dynamic>? meta,
}) {
  return registry.register(
    id,
    ElementHandler(
      type: ElementType.toggle,
      handler: handler,
      getValue: getValue,
      disabled: disabled,
      hint: hint,
      meta: meta,
    ),
  );
}
